abi-versions/README.md

```diff
@@ -0,0 +1,40 @@
+# Proxy-Wasm ABI specification
+
+## ABI versioning
+
+The ABI is in the form of: `<major>.<minor>.<patch>`
+
+Breaking changes require increasing the major version.
+
+Adding new functions requires increasing the minor version.
+
+Adding new enum values requires increasing the patch version.
+
+Wasm modules should export `proxy_abi_version_<major>_<minor>_<patch>` function, indicating which
+version of the ABI they were compiled against.
+
+Host environments should be able to support multiple versions of the ABI.
+
+## Naming conventions
+
+All functions are using `snake_case`, and they start with the `proxy_` prefix. Implementation
+specific extensions should use different prefixes (e.g. `envoy_`, `nginx_`, `ats_`, etc.) in order
+to avoid polluting the common namespace. However, this is highly discouraged, and should be used
+only for implementing optional proxy-specific functionality, otherwise we’re going to lose
+portability.
+
+Function names are in the form of: `proxy_<verb>_<namespace>_<field>_<field>` (e.g.
+`proxy_on_http_request_headers`, `proxy_get_map_value`, etc).
+
+## Memory ownership
+
+No memory-managed data is passed to Wasm module in any of the event callbacks, and Wasm module must
+request memory-managed data from the host environment if it’s interested in it, e.g.
+`proxy_on_http_request_body` only passes the information about the number of available request body
+bytes, but the module must request them using `proxy_get_buffer`. When that happens, the host
+environment will request the Wasm module to allocate memory using `proxy_on_memory_allocate` and
+copy requested data into it. In doing so, it passes memory ownership of that data to the Wasm
```

Comment:
This seems unfortunate if I want to write a high-performance WASM filter that examines or modifies content.

What do you you think of supporting, in the context of proxy_on_http_request_body, some APIs to enable a filter to access the body in various ways without copying all of it? E.g.

 * copying ranges of bytes into the WASM address space
 * searching for bytes/substrings
 * global-replace
 * global-replace-RE2
 * parse as JSON and provide JSON API to content
 * send content to a socket without brining it into the WASM space

Maybe not all of these should be done at first (or ever) but establishing this pattern seems useful if you want to write performance content-aware WASM filters.

Reply:
> * copying ranges of bytes into the WASM address space

`proxy_get_buffer` takes `offset` and `max_size` as a parameter, so this is already supported.

> * searching for bytes/substrings
> * global-replace
> * global-replace-RE2
> * parse as JSON and provide JSON API to content

While those APIs could be implemented, they seem very much like a stretch. We don't have those implemented in `BufferImpl`, for example. Also, it's unclear why those couldn't be implemented inside Wasm module.

> * send content to a socket without brining it into the WASM space

I'm not sure I follow. Where would this content came from? What's the use case?

> I am curious if future WASM VMs or evolutions will support zero copy from outside.

We very much want to support zero-copy, and as soon as Wasm gets support for multiple-memories that can be "mounted" at runtime, we're going to add support for it.

Reply:
proxy_get_buffer sgtm.

The other APIs could help users that want to implement transformations that do not require copying the entire response-body.

The "send content to a socket" could be used for proxying requests to an external transform-server, without the extra copy. One example (which might not be practical as there's already a native impl) is a gzip encoder or jpeg optimizer running out-of-process.

RE wasm support for multiple memories: I don't fully understand what that means, but do you if that's a specific plan from Wasm maintainers?

Reply:
Also... WDYT about streaming? Say there's a long chunked response; can WASM wake up and process chunks as they come in? Or does the WASM engine not get to look at the first byte until the last byte has arrived?

Apologies if this was already covered.

Reply:
I implemented CEL expression foreign function call for this exact reason. It's very expensive to copy data just to do a little computation.